src/frontend/src/scheduler/distributed/stage.rs

```diff
@@ -672,6 +718,19 @@ impl StageRunner {
             .await;
     }
 
+    async fn notify_stage_state_changed<F>(&self, new_state: F, msg: QueryMessage)
```

Comment:
How about simply

```suggestion
    async fn notify_stage_state_changed<F>(&self, old_state: StageState, new_state: StageState, msg: QueryMessage)
```

Reply:
No, this way we can't do the check in passed closure.

Reply:
I mean

```rust
async fn notify_stage_state_changed<F>(
    &self,
    old_state: StageState,
    new_state: StageState,
    msg: QueryMessage,
) where
    F: FnOnce(StageState) -> StageState,
{
    let mut s = self.state.write().await;
    let r = mem::replace(&mut *s, new_state);
    assert_matches!(r, old_state);

    self.send_event(msg).await;
}
```

and

```rust
self.notify_stage_state_changed(
    StageState::Started,
    StageState::Running,
    QueryMessage::Stage(StageEvent::Scheduled(self.stage.id)),
)
```

What's wrong with this?

Reply:
This signature is more limited than current approach, what if the expected old state is more than 1?

Reply:
I think this is a small style problem and this is blocking the memory control policy for yuanxin and other fixes, how about we merge it first and do it later if it's proven to be necessary? cc @fuyufjh

Reply:
I was not convinced but you didn't give me a choice. 🙁